graph.go

```diff
@@ -0,0 +1,293 @@
+)
+
+func (g *Graph) Last() *Spec {
+	return g.ordered[len(g.ordered)-1]
+}
+
+func (g *Graph) Lock() {
+	graphLock.Lock()
+	return
+}
+func (g *Graph) Unlock() {
+	graphLock.Unlock()
+	return
+}
+
+func InitGraph(specs []*Spec, subTarget string) error {
+	if BuildGraph != nil {
+		return nil
+	}
+
+	if BuildGraph == nil {
+		BuildGraph = new(Graph)
+		BuildGraph.Lock()
+		defer BuildGraph.Unlock()
+		*BuildGraph = Graph{
+			target:   subTarget,
+			edges:    sets.New[dependency](),
+			vertices: make([]*vertex, len(specs)),
+			specs:    make(map[string]*Spec),
+			indices:  make(map[string]int),
+		}
+	}
+
+	if len(specs) == 1 {
+		BuildGraph.target = specs[0].Name
+	}
+
+	for i, spec := range specs {
+		name := spec.Name
+		BuildGraph.specs[name] = spec
+		v := &vertex{name: name}
+		BuildGraph.indices[name] = i
+		BuildGraph.vertices[i] = v
+	}
+
+	for name, spec := range BuildGraph.specs {
+		if spec.Dependencies == nil {
+			continue
+		}
+		vi := BuildGraph.indices[name]
+		v := BuildGraph.vertices[vi]
+		type depMap struct {
+			kind string
+			m    map[string][]string
+		}
+		runtimeAndBuildDeps := []depMap{
+			{m: spec.Dependencies.Build, kind: "build"},
+			{m: spec.Dependencies.Runtime, kind: "runtime"},
+		}
+		for _, deps := range runtimeAndBuildDeps {
+			if deps.m == nil {
+				continue
+			}
+			for dep, constraints := range deps.m {
+				_ = constraints // TODO(pmengelbert)
+				if name == dep {
+					continue // ignore if cycle is length 1
+				}
+				wi, ok := BuildGraph.indices[dep]
+				if !ok {
+					// this is dependency in the package repo
+					continue
+				}
+				w := BuildGraph.vertices[wi]
+				BuildGraph.edges.Insert(dependency{
+					v:    v,
+					w:    w,
+					kind: deps.kind,
+				})
+			}
+		}
+	}
+
+	return BuildGraph.topSort()
+}
+
+// https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+func (g *Graph) topSort() error {
+	index := 0
+
+	s := stack.New[*vertex]()
+	output := cycleList{}
+
+	var strongConnect func(v *vertex) error
+	strongConnect = func(v *vertex) error {
+		v.index = new(int)
+		*v.index = index
+		v.lowlink = index
+		index++
+
+		s.Push(v)
+		v.onStack = true
+
+		for edge := range g.edges {
+			if v.name != edge.v.name {
+				continue
+			}
+
+			w := edge.w
+			if w.index == nil {
+				if err := strongConnect(w); err != nil {
+					return err
+				}
+
+				v.lowlink = minimum(v.lowlink, v.lowlink)
+				continue
+			}
+
+			if w.onStack {
+				v.lowlink = minimum(v.lowlink, *w.index)
+			}
+		}
+
+		if v.lowlink == *v.index {
+			component := []*vertex{}
+
+			var w *vertex
+			isSome := func(o stack.Option[*vertex]) bool {
+				if o.IsSome() {
+					w = o.Unwrap()
+					return true
+				}
+				return false
+			}
+
+			for opt := s.Pop(); isSome(opt); opt = s.Pop() {
+				w.onStack = false
+				component = append(component, w)
+				if w == v {
+					break
+				}
+			}
+
+			w.onStack = false
+			output = append(output, component)
+		}
+
+		return nil
```

Comment:
looks like this returns nil in all cases? do we need to return an error from this function?

Reply:
done